feat(training)!: Configurable weight averaging (exponential moving average (EMA), ...)

jakob-schloer opened the pull request:
Weight averaging is a technique to improve model generalization by averaging model weights during training. In the new pytorch lightning version (2.6.0) different versions of weight averaging are now supported. 

Here, I refactor the existing SWA implementation to support general weight averaging methods through the config. For example, EMA training can be used by adding the following to the `config.training`: 
```
         weight_averaging:
            _target_: pytorch_lightning.callbacks.EMAWeightAveraging
            decay: 0.999
```

The impact of EMA for smoother training and less fluctuation between checkpoints is especially important for the biases at longer rollout: 

<img width="546" height="297" alt="image" src="https://github.com/user-attachments/assets/5cd80788-1ee6-4dce-ac06-31261592872f" />

Figure: Validation loss over training iterations for 2t. Upper panel shows validation loss at step 1 (1 day) and lower panel step 14 (14 days). 



**Note:** 
- I've deliberately deleted the previous version of weight initialization to keep the code cleaner. 
- Old configs should still work, since defaults are null if `weight_averaging` does not exist, but this is still a breaking change if swa was used before.



***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)


<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--743.org.readthedocs.build/en/743/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--743.org.readthedocs.build/en/743/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--743.org.readthedocs.build/en/743/

<!-- readthedocs-preview anemoi-models end -->

jakob-schloer committed:
Configurable weight initialization with lightning>=2.6.

anaprietonem commented:
@jakob-schloer  - How you're getting on with this PR? Are you planning to work on this soon?

jakob-schloer committed:
Fix EMA to work with dynamic scalars.

jakob-schloer committed:
Migration script for SWA to weight averaging.

jakob-schloer committed:
Fix configs.

anaprietonem commented on models/src/anemoi/models/migrations/scripts/1776237003_rename_swa_to_weight_averaging.py:
```diff
@@ -0,0 +1,72 @@
+    return ckpt
+
+
+def rollback(ckpt: CkptType) -> CkptType:
```
This functionality was deprecated, so migration checkpoints do not need to provide this now. You can remove the rollback function

jakob-schloer replied:
Good to know. I've removed the rollback function.

anaprietonem commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -58,6 +57,96 @@ def nestedget(config: DictConfig, key: str, default: Any) -> Any:
+        current_buf.data.copy_(avg_buf.data)
+
+
+def _get_weight_averaging_callback(config: DictConfig) -> list[Callback]:
```
I'd suggest to move those to it's own file and then just import this function in the callbacks __init__ so that we don't fill the __init__ with function related to this

jakob-schloer replied:
Good idea. I've created a `weight_averaging.py`.

anaprietonem commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -58,6 +57,96 @@ def nestedget(config: DictConfig, key: str, default: Any) -> Any:
+
+    callback = instantiate(weight_averaging_config)
+
+    # Patch swap/copy methods to use name-based matching. Needed for dynamic scalers like NaNMaskScaler.
```
Jakob could you explain a bit more in detail why this is needed? and when/where it's used? Have to tested this when sharding the model?

jakob-schloer replied:
The patch swap method is needed because the lightning `AveragedModel` uses positional pairing but our `ScaleTensors` are stored as `register_buffer(name, scaler, persistent=False)`. So once the AverageModel has been created and a dynamical scaler later adds/removes a buffer they no longer line up in position. 
The `_safe_swap_models` and `_safe_copy_average_to_current` pair entries by name instead of position.

anaprietonem commented on training/src/anemoi/training/losses/scaler_tensor.py:
```diff
@@ -336,7 +336,7 @@ def update_scaler(self, name: str, scaler: torch.Tensor, *, override: bool = Fal
 
         original_scaler = self._tensors.pop(name)
-        original_scaler_buffer = self._buffers.pop(name, None)
+        original_scaler_buffer = self._buffers.get(name)
```
why is this change required?

jakob-schloer replied:
It's part of the same problem as above with the dynamic scalers. `_buffers` is an `OrderedDict`. If you pop and later call `register_buffer` again, the buffer gets inserted at the end. So every time a dynamic scaler updated itself, that buffer jumped to the end of the iteration order. So over the course of training, `named_buffers()` kept reshuffling.
We need both fixes because the pop -> get change fixes the EMA "live module" (`AveragedModel.update_parameters`) buffer during training and the `_safe_swap_models` and `_safe_copy_average_to_current` fix ensures that the buffers of the EMA "averaged module" and the "live module" matches.

anaprietonem commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -58,6 +57,96 @@ def nestedget(config: DictConfig, key: str, default: Any) -> Any:
+
+    Example config:
+        weight_averaging:
+          _target_: anemoi.training.diagnostics.callbacks.optimiser.ExponentialMovingAverage
```
I think this example needs updating, to use pytorch_lightning.callbacks.EMAWeightAveragin right?

jakob-schloer replied:
Yes, changed that. I had a version with my own EMA class before :)

anaprietonem commented on training/src/anemoi/training/config/training/diffusion.yaml:
```diff
@@ -39,11 +39,14 @@ gradient_clip:
-swa:
-  enabled: False
-  lr: 1.e-4
+# weight averaging only supported in Lightning 2.6+
```
is this both for EMA and SWA? If so we should add a check for the torch version in the code before those gets instantiated so the error is properly caught.

jakob-schloer replied:
Sure, I've added the check for lightning>=2.6.0.

anaprietonem commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -58,6 +57,96 @@ def nestedget(config: DictConfig, key: str, default: Any) -> Any:
+    if weight_averaging_config is None:
+        return []
+    if not isinstance(weight_averaging_config, dict | DictConfig):
+        return []
```
if this is returning an empty list, then why not using in the config as default: 
weight_averaging: [ ] as default in the config for consistency?

it's also not super clear why we return a list here, if one should just be able to use just one WA technique right?

jakob-schloer replied:
Every `_get_*_callback` method returns a list and the caller in the `__init__.py` extends them. I've just sticked to the convention there. 
Changing the default to an empty list would suggest that multiple weight averaging techniques made sense - which they don't. 
For those two reasons, I leave the code as it is.

matschreiner commented:
Hi Jacob! Looks really nice, made a small PR with my suggestions from the ATS metting - https://github.com/ecmwf/anemoi-core/pull/1047

anaprietonem commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -58,6 +57,96 @@ def nestedget(config: DictConfig, key: str, default: Any) -> Any:
+    if not isinstance(weight_averaging_config, dict | DictConfig):
+        return []
+    if "_target_" not in weight_averaging_config:
+        return []
```
I'd suggest to add some logging statements clear to help debug if the callback it's getting loaded correctly or getting skipped

jakob-schloer replied:
Sure, I've added some: 
```
  if weight_averaging_config is None:
        LOGGER.debug("No weight averaging configured. Skipping.")
        return []
    if not isinstance(weight_averaging_config, dict | DictConfig):
        LOGGER.warning(
            "training.weight_averaging has unexpected type %s; expected a dict with '_target_'. Skipping.",
            type(weight_averaging_config).__name__,
        )
        return []
    if "_target_" not in weight_averaging_config:
        LOGGER.warning("training.weight_averaging is set but has no '_target_' field. Skipping.")
        return []

    callback = instantiate(weight_averaging_config)
    LOGGER.info("Loaded weight averaging callback: %s", weight_averaging_config["_target_"])
```

anaprietonem commented on training/src/anemoi/training/schemas/training.py:
```diff
@@ -66,16 +66,42 @@ class GradientClip(BaseModel):
-    lr: NonNegativeFloat = Field(example=1.0e-4)
-    "Learning rate for SWA."
+    target_: Literal[
+        "pytorch_lightning.callbacks.EMAWeightAveraging",
```
Regarding those schemas, since we are using pytorch_lightning classes directly, my recommendation would be to keep these consistent with the OptimizerSchema (lenient one) so that then we can pass as many args as we need and we rely on the API defined by lightning, and do not risk getting differences in the defaults with validation on/off that we saw before for the optimiser.

In the future if we have a custom anemoi WA class that we build on top of any of the PTL ones, then that one should include a schema.

jakob-schloer replied:
That's a much better solution.

jakob-schloer committed:
Address PR comments.

matschreiner committed:
fix: update weight averaging config to not have duplicated commented out code.  (#1047)

Hi Jacob, here is what I meant in the ATS :)

 - Add training/weight_averaging/ema.yaml and swa.yaml
- Add - weight_averaging: null to defaults in all training configs
- Remove inline weight_averaging: null and commented-out example configs
  - Can now be selected via CLI:
  training/weight_averaging=ema or
  training/weight_averaging=swa

---------

Co-authored-by: Jakob Schloer <jakob.schloer@gmail.com>

jakob-schloer commented on training/src/anemoi/training/config/training/weight_averaging/ema.yaml:
```diff
@@ -0,0 +1,3 @@
+_target_: pytorch_lightning.callbacks.EMAWeightAveraging
+decay: 0.999
+update_starting_at_step: ${training.lr.warmup}
```
```suggestion
```

jakob-schloer replied:
Refactor of optimizers would make that error prone.

jakob-schloer committed:
Remove linkto learning rate warmup.

jakob-schloer commented:
Test with model sharding: 
https://mlflow.ecmwf.int/#/experiments/413/runs/cedd76dd82974a3ba3423b62b5974381/artifacts

Integration test:
https://github.com/ecmwf/anemoi-core/actions/runs/24857081535

anaprietonem reviewed (approved):
Thanks Jakob for addressing the comments and doing those extra checks! PR LGTM, feel free to merge!

jakob-schloer committed:
Merge branch 'main' into feature/configurable_weight_averaging
